### ISSUE TITLE
Optimize editor for input latency

### ISSUE THREAD

**Issue by Tyriar** [microsoft/vscode#161622]
I recently did an exploration into input latency and found it can get pretty bad on slower machines. A lot of the problem is related to how we use synchronous event emitters/listeners work, performing their work after keypress but before the animation frame.

![image](https://user-images.githubusercontent.com/2193314/192010544-dc605955-2e0b-4e4a-8ed7-5619fe69ee2e.png)

My proposal to improve is:

1. Review the most time consuming event listeners. If they are clearly safe to be performed at a later time without worrying about race conditions being introduced., move them to use an asynchronous event listener to be performed after the animation frame. For example UI updates like activity bar badge and tab indicator when an editor becomes dirty
2. Review all events and their listeners, defer as much as possible to after the animation frame. One way of doing this is for each important event to have both a sync and an async event. It's unclear how much we can move exactly, but when we do this we need to be extremely careful to not introduce text buffer-related race conditions as there are some assumptions made that we may be breaking by doing this
3. Setup development tools and/or telemetry to easily track measuring latency, I created https://github.com/microsoft/vscode/tree/tyriar/measure_latency to demonstrate a technique to approximate input latency
4. Come up with a plan for how we can prevent regressions for this critical path code

Tentatively assigning to October

**Comment by bpasero** [comment]
The core listener for text editors to react on content changes is:
* https://github.com/microsoft/vscode/blob/da506e9b8d86a7265d53255ef8a8603d792d74f5/src/vs/workbench/services/textfile/common/textFileEditorModel.ts#L565
* https://github.com/microsoft/vscode/blob/da506e9b8d86a7265d53255ef8a8603d792d74f5/src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts#L361

These drive a ton of things on top such as:
* dirty state indication throughout the UI (only when dirty state changes, not on every content change)
* editor auto save
* editor backups
* etc (anyone using `onDidChangeDirty` and friends)

I wonder how an async emitter would help here: yes, it would take away lag from the first character typing when the editor transitions into being dirty, but eventually we have to pay the price, so the lag would just happen later? Or is the idea to delay the event literally on idle time?

**Comment by Tyriar** [comment]
@bpasero oh I missed the question there. The idea is to delay it until shortly after via setTimeout, so the text change should appear asap and the dirty indicator (as an example) would appear 1 or 2 frames later. ie:

Current:

- keypress task:
  - handle dirty change
  - schedule various things (backup, auto save, etc.)
  - other less critical updates (bracket pair parsing?)
  - render

Desired:

- keypress task:
  - handle dirty change
  - render
- following task:
  - schedule various things (backup, auto save, etc.)
  - other less critical updates (bracket pair parsing?)
  - render (whatever changed in this task)

There's some nuance here in what should be handled in the keypress task. For example currently the suggest widget is moved and updated in the keypress task. What we probably want is for the suggest widget to move in the keypress task but defer updating as it can be very expensive. Things like this we'll need to experiment with to see if splitting it up ends up with a worse UX and should be on the critical path.

**Comment by Tyriar** [comment]
Realized one of my laptops has a CPU similar to the average users (though a better GPU). Here's a screenshot of typing `this` into `TerminalInstance.ctor` which validates my assumptions that latency is pretty terrible on lower hardware (up to 100ms in this case):

![image](https://user-images.githubusercontent.com/2193314/195095702-d62a05f2-53c2-4110-baeb-11b64822d499.png)

This is with a i7-8750H @ 2.2GHz, didn't turn off turbo boost which can push it up to 4.10GHz. Not entirely sure how that works but I think I can disable it in BIOS if needed.

**Comment by Tyriar** [comment]
I drilled into a profile on my macbook to understand some parts a little more. Here are the details

TLDR: An enormous amount of work seems to be spent just scheduling things, `Event.defer` will probably be an easy solution to those.

![image](https://user-images.githubusercontent.com/2193314/195705533-7b8abffc-b345-4780-ad17-0d2ca80142b8.png)

## Latency

- Keydown to character on screen ~30.52ms
- Keydown to suggest on screen ~106.6ms

## High level parts

- Critical path 30.52ms
  - Key down 1.85ms (6%)
  - Key press 15.43ms (51%) 
  - Render animation frame 6.91ms (23%)
  - Render to composite 4.56ms (15%)
- Suggest
  - Waiting for suggestions from extension host 37.1ms
  - Suggest widget setup/adding 30.87ms
  - Suggest widget rendering 8.37ms

## Key press

15.43ms (51% of critical path)

### Most expensive bottom up parts

- setTimeout 3.4ms (22%)
  - Async timeouts are definitely disabled
  - A lot of this is due to the sheer amount of timeouts, a shared timeout like in Event.defer will improve
  ![image](https://user-images.githubusercontent.com/2193314/195705411-6f7cc856-9b96-4ed6-b893-6d55b5357d57.png)
- Recalculate style and layout 1.8ms (12%)
  - This is done because of the text input event from the keypress, nothing we can do if we're using an <input>/<textarea>
- clearTimeout 0.7ms (5%)
  - We may be able to avoid clearing some?
- requestAnimationFrame 0.2ms (1.5%)
- requestIdleCallback 0.2ms (1.5%)
- setInterval 0.2ms (1.5%)

### What is it doing?

Legend:

:bulb: We can probably improve or defer this fairly easily
:question: We can maybe improve this, needs more investigation/validation

- _type @ codeEditorWidget.ts 11.54ms
  - _executeCursorEdit 10.57ms (69%)
    - _executeEdit 2.99ms
      - pushEditOperation @ editStack.ts 0.96ms
        - :bulb: parseDocumentFromTextBuffer @ bracketPairsTree.ts 0.12ms (can be merged with below occurrence)
      - endDeferredEmit
        - onDidChangeContentOrInjectedText @ viewModelImpl.ts 0.39ms
        - modelService scheduling 0.16ms
          - :bulb: Consolidate scheduling?
        - scheduleBackup 0.5ms
          - :bulb: Cancels and restarts a timeout
        - triggerDiff @ dirtyDiffDecorator.ts 0.13ms
          - :bulb: Consolidate scheduling?
        - Schedule syncing with ext host 0.5ms
          - :bulb: Stringify and multiple setTimeouts
          - This is already async, perfect for Event.defer
      - setSelections @ cursorCollection.ts 0.11ms
    - endEmitViewEvents 
      - _emitMany @ viewModelEventDispatcher.ts
        - _scheduleRender @ view.ts 0.24ms
          - :question: Mainly just requests an animation frame
        - writeScreenReaderContent @ textAreaInput.ts 1.41ms
          - This is a result of setting value on textarea, triggers a recalc style, layout, scroll which is a little weird?
          - :bulb: Defer to after render when accessibility mode is off?
        - onLinesChanged @ viewLayer.ts 0.13ms
        - onCursorStateChanged 0.62ms
          - :question: Updates the cursor blinking; on input the interval gets reset
      - _emitOutgoingEvents @ viewModelEventDispatcher.ts
        - :bulb: Bracket pair matching scheduling 0.11ms
        - computeLinks @ links.ts0.35ms
          - _updateScores @ languageFeatureRegistry.ts 0.11ms
            - :bulb: Schedules language detection?
          - :bulb: More unknown scheduling 0.24ms
        - Schedule tokenize viewport 0.11ms
          - :question: This is done even though tokenize if cheap runs?
        - :bulb: Schedule multicursor selection highlighter 0.13ms
        - :bulb: Schedule updateInlineValuesScheduler @ debugEditorContribution.ts 0.38ms
        - notifyNavigation @ historyService.ts 0.13ms
          - This is keeping the history stack up to date, probably needed
        - triggerFolderingModelChanged @ folding.ts 0.24ms
          - :bulb: Schedules folding changes
          - :bulb: This creates a new function each trigger? Extra time is probably spent in V8 compiling/optimising
        - beginComputeCommentingRanges @ commentsEditorContribution.ts 0.11ms
          - :bulb: Just schedules
        - onDidChangeModelContent @ lightBulbWidget.ts 0.12ms
          - :bulb: This just gets the editor model and hides the light bulb widget if needed, good candidate for deferral
        - unicodeHighlighter.ts 0.13ms
          - :bulb: Just schedules
        - inlayHintsController.ts 0.38ms
          - :bulb: Just schedules and debounces for 1250ms
        - documentSymbolsOutline.ts 0.11ms
          - :bulb: Just schedules/debounces
        - _onModelChange @ codeLensController.ts 2.87ms
          - :bulb: This is a lot of time devoted to code lenses, could all these be deferred?
          - changeDecorations @ codeEditorWidget.ts 2.74ms
            - Creating a changeAccessor object ~0.15ms
              – :bulb: Could be optimized
          - endDeferredEmit 2.61ms
            - languageDetection.contribution.ts 0.37ms
              – :bulb: Just scheduling
            - inlineCompletionsModel.ts 0.25ms
              – Update ranges, updateFilteredInlineCompletions
              – :bulb: Schedule auto update
            - bracketMatching.ts 0.40ms
              – :bulb: Just scheduling for 50ms
            - suggestWidgetPreviewModel.ts 0.61ms
              – :question: Updates and schedules
            - onSelectionChange @ editorStatus.ts 0.13ms
              – :question: Update
            - _onCursorChange @ codeActionModel.ts 0.25ms
              – :bulb: Just scheduling
            - inlineCompletionsModel.ts 0.13ms
              – :question: Updating
              – Is this doing duplicate work as above? Can this be debounced to a microtask?
            - onDidChangeCursorPosition @ documentSymbolsOutline.ts 0.11ms
              – :bulb: Just scheduling
            - _onCursorChange @ suggestModel.ts 0.26ms
              – :question: Schedules based on some state and the selection
            - mainThreadEditor.ts 0.11ms
              – :question: Generates a delta of the editor and fires properties change (to sync with ext host?)
              – :question: ` _readVisibleRangesFromCodeEditor` which calls codeEditor.getVisibleRanges() is the most expensive thing here
          - Schedule asking for all references 0.13ms
            - Not sure we could defer this but we could probably optimize
  - _type @ codeEditorWidget.ts 0.97ms
    - onDidType @ parameterHintsModel.ts 0.1ms
    - handleUserInput  @inlineCompletionsModel.ts 0.14ms
      - Just hides or schedules
    - checkTriggerCharacter @ suggestModel.ts 0.73ms
      - tokenizeIfCheap 0.49ms
        - parseDocumentFromTextBuffer @ bracketPairsTree.ts 0.13ms (! This is much worse on my Windows machine?)
      - provideSuggestionItems
- Microtasks 0.31ms
  - ? @ defaultWorkerFactory.ts 0.15ms
    - Not sure what triggers this

**Comment by Tyriar** [comment]
The methodology, using the "Performance" tab was flawed, it exagerates timeouts (even when async stacks are disabled) and the overhead slows it by around 4x. A profile in the JavaScript Profiler tab is much more accurate and you can also perform the operation many times and then look as the slow parts in terms of the percentage of time they take up. By looking at the percentage instead of actual milliseconds it's much more reliable to get a good sample as it consolidates all calls.

For example I recorded typing many bunch of characters, in top down you can now see a more accurate view of the function. This `fn` is the keypress handler and it shows it took up 17.85% of the CPU time:

<img width="426" alt="Screen Shot 2022-10-14 at 8 38 51 am" src="https://user-images.githubusercontent.com/2193314/195886303-68dd0d8b-d621-44e9-9227-acc634607c47.png">

Now focusing on that function shows the breakdown of the call stack as a percentage of the total `fn` time. One of the suspected functions contributing to the slow down is the bracket pair parsing which you can see takes up 1% of the time after cheap tokenization:

<img width="924" alt="Screen Shot 2022-10-14 at 8 40 50 am" src="https://user-images.githubusercontent.com/2193314/195886694-e109c59b-211d-4d6a-9ee7-ca799a0b749b.png">

And 8.39% of the time after the model content changes:

<img width="934" alt="Screen Shot 2022-10-14 at 8 43 14 am" src="https://user-images.githubusercontent.com/2193314/195887201-d5a5dae4-da11-4ea6-822c-b25c6973d6e1.png">

Looking into this some more now.

**Comment by Tyriar** [comment]
Here's a breakdown of the CPU profile when typing a bunch of characters (~50, random), the suggest widget didn't show most of the time. This tree isn't complete, a lot is omitted here to remove noise vs just looking at the profile in devtools. The children of a node are not ordered and not necessarily directly below the parent or siblings of other children.

This is a living document for now:

- 100%: fn - keypress handler
  - 94.53%: _type
    - 11.55%: onWillType
      - 10.75%: checkTriggerCharacter - suggestModel.ts
        - 0.30%: provideSuggestionItems - suggest.ts
        - 0.68%: getWordAtPosition - textModel.ts
        - 9.30%: tokenizeIfCheap - tokenizationTextModelPart.ts
    - 82.76%: type
      - 52.68%: endEmitViewEvents - viewModelEventDispatcher.ts
        - 44.59%: emitOutgoingEvents
          - onDidChangeModelContent
            - 0.87%: editorStatus.ts
            - 27.16%: suggestModel.ts - refilters items
              - 21.12% showSuggestions - suggestWidget.ts
                - [ ] Drill into this
              - 2.51%: cancel - suggestModel.ts
                - [ ] 2.51% to cancel?
              - 1.33%: shouldAutoTrigger - suggestModel.ts
            - 0.23%: inlayHintsController.ts
            - 0.27%: clickLinkGesture.ts
            - 0.57%: links.ts
            - 0.34%: unicodeHighlighter.ts
            - 0.30%: commentsEditorContribution.ts
            - 1.67%: lightBulbWidget.ts
              - 1.37%: _updateLightBulbTitleAndIcon
                - [ ] Defer DOM interaction?
            - 1.56%: textEditor.ts
              - 1.52%: handleActiveEditorSelectionChangeEvent - historyService.ts - Records navigation for back/forward
                - [ ] Defer?
            - 0.27%: viewPortSemanticTokens.ts
            - 0.57%: folding.ts
              - [ ] This throws away and re-creates a lot of stuff
            - 0.27%: documentSymbolsOutline.ts
              - [ ] Does sharing a timeout help?
            - 8.58%: codelensController.ts
              - 8.36%: changeDecorations - textModel.ts
                - 0.38%: emitMany - viewModelEventDispatcher.ts
                - 7.44%: _emitOutgoingEvents - viewModelEventDispatcher.ts
                  - 0.23%: multicursor.ts
                  - 0.84%: suggestWidgetPreviewModel.ts
                  - 0.46%: inlineCompletionsModel.ts
                  - 0.61%: textEditor.ts
                    - 0.57%: historyService.ts - records navigation
                      - [ ] This happens multiple times
                  - 0.27%: suggestModel.ts
                  - 1.41%: editorStatus.ts - Updating the status bar lines/cols
                  - 0.27%: languageDetectionContribution.ts
                  - 0.15%: bracketMatching.ts
                  - 0.57%: inlineCompletionsModel.ts
                  - 1.71%: mainThreadEditor.ts
                    - [ ] More syncing with ext host?
                  - 0.19%: codeActionModel.ts
            - 0.49%: testingDecorations.ts
        - 8.01%: emitMany
          - 6.96%: handleEvents - viewEventHandler.ts
            - 0.68%: onRevealRangeRequest - viewLines.ts
            - 0.84%: onCursorStateChange - viewCursors.ts
            - 0.19%: onCursorStateChange - currentLineHighlight.ts
            - 4.18%: onCursorStateChange - textAreaHandler.ts - writeScreenReaderContent
              - [ ] Defer writing screen reader content when accessibility support is disabled
          - 0.72%: handleEvents - view.ts
            - 0.46%: _scheduleRender - view.ts
              - [ ] Investigate
      - 30.08%: type
        - 25.29%: _executeEditOperation
          - 23.59%: pushEditOperations
            - 14.28%: _pushEditOperations
              - 1.71%: applyEdits
              - 0.49%: acceptEdit
              - 0.76%: acceptReplace
              - 8.93%: _emitContentChangedEvent
                - 8.05%: parseDocumentFromTextBuffer - bracketPairsTree.ts
                - 0.42%: _beginBackgroundTokenization
            - 8.51%: endDeferredEmit
              - 1.60%: onDidChangeContentOrInjectedText - viewModelImpl.ts
              - 6.42%: onDidChangeContent - textModel.ts
                - 3.04%: onDidChangeContent - mainThreadDocuments.ts - Syncing model changes with ext host
                - 0.42%: onDidChangeContent - dirtydiffDecorator.ts
                - 2.70%: onDidChangeContent - textFileEditorModel.ts
                  - 1.82%: onDidChangeContent - workingCopyService.ts
                    - 0.87%: workingCopyHistoryTracker.ts
                    - 0.76%: workingCopyBackupTracker.ts

**Comment by Tyriar** [comment]
Profile when typing `.` after this and backspace repeatedly to measure high-level impact of suggest widget.

- 96.19% `_type` - codeEditorWidget.ts
  - 5.32% `checkTriggerCharacter` suggestModel.ts 
  - 72.37% `_insertSuggestion` - suggestWidget.ts
    - 63.53% `insert` - snippetController2.ts
	  - [ ] Snippet leading to show suggest?
	  - 61.15% `insert` - snippetSession.ts
	    - 58.85% `executeEdits` - codeEditorWidget.ts
      - [ ] What's being executed? Is this just in an unexpected spot in the stack trace?
      - 48.99% `endEmitViewEvents`
        - 43.74% `_refilterCompletionItems` - suggestModel.ts
    	  - 43.31 _onNewContext (firing event) - suggestModel.ts
    	    - 42.73% `showSuggestions` - suggestWidget.ts
          - 37.12% `_layout` - suggestWidget.ts
            - 30.00% `getScrolledVisiblePosition` codeEditorWidget.ts
        	  - [ ] No scrolling was happening in either the editor or the suggest widget
        	  - 29.71% `_renderNow` - view.ts
            - 5.68% render - minimap.ts
              - [ ] Minimap shouldn't block text rendering?
            - 4.60% render - contentWidgets.ts
              - 4.6% afterRender - suggestWidget.ts
            	- [ ] Investigate
            - 2.01% prepareRender - viewOverlays.ts
              - 1.44% prepareRender - indentGuides.ts
            	- [ ] Investigate
            - 7.34% prepareRender - textAreaHandler.ts
              - 7.34% _readPixelOffset - viewLine.ts
                - [ ] DOM access?
            - 5.54% renderText - viewLines.ts
              - [ ] This is happening on keypress as well as animation frame?
            - 5.61% getDomNodePagePosition - dom.ts
        	  - [ ] DOM access?
          - 4.96% `splice` - listWidget.ts
        - 9.50% `(`anonymous) - viewModelImpl.ts
    	  - 8.63% `pushEditOperations` - TextModel.ts
    - 1.08% `memorize` - suggestMemory.ts
    - 6.12% `cancel` - suggestModel.ts
	  - 6.04% `hideWidget` - suggestWidget.ts
	    - [ ] Hiding is 6% when it shouldn't be getting hidden?
    - 3.02% `splice` - listWidget.ts
  - 17.77% `type` - viewModelImpl.ts
    - Similar to the profile above
    - 4.46% `typeWithInterceptors` - cursorTypeOperations.ts
      - 4.17% `_runAutoIndentType` - cursorTypeOperations.ts
    - 6.04%_`executeEditOperation`

**Comment by Tyriar** [comment]
The big revelation above is that there's a full render in both the keypress and the animation frame events, triggered by the suggest widget 🤯:

![image](https://user-images.githubusercontent.com/2193314/196268849-7ad8cf37-0a20-416c-b720-5471b5baef89.png)

**Comment by Tyriar** [comment]
Here's a breakdown of the list `splice` calls made when filtering suggest. This is a general area I've noticed is a little slow in other areas like search. This particular profile ("c", backspace, repeat to refilter on both) showed `ListWidget.splice` taking up 11.28% of the total keypress task.

![image](https://user-images.githubusercontent.com/2193314/196492047-dd2d51db-e727-4032-9212-26e3251d62b1.png)

**Comment by Tyriar** [comment]
After pulling all the proposed changes into a single branch this is what pressing `.` after a `this` looks like. The red will mostly go away if we implemented the GPU-based renderer (https://github.com/microsoft/vscode/issues/162445), the blue part is minimap rendering which could be significantly sped up by moving to webgl.

![image](https://user-images.githubusercontent.com/2193314/196552761-fd6b2dbe-4e0e-4448-b1b5-4f48b0a3dbe2.png)

Current main: <details>
![image](https://user-images.githubusercontent.com/2193314/196554587-28bf3030-3a8e-4c47-a069-74cc1c7b3621.png)
</details>


After the completion response comes back from the exthost this happens:

![image](https://user-images.githubusercontent.com/2193314/196553613-d7aad928-d44d-42d7-b624-d4cc6caa5a34.png)

Current main: <details>
![image](https://user-images.githubusercontent.com/2193314/196554707-42778e92-a70a-48a8-8d44-abce2b5fa67d.png)
</details>

I couldn't find many more wins here, most of it is bound by list splice and fetching the position of elements from the DOM (maybe something here could improve?). There are wins from moving to webgl as well but it would probably be more effort than it's worth to maintain.

---

Here is a look at re-filtering the already visible suggest widget:

![image](https://user-images.githubusercontent.com/2193314/196553995-0e57a40d-7f54-457d-8e87-a9d9839869d6.png)

Current main: <details>
![image](https://user-images.githubusercontent.com/2193314/196554790-232c5456-28b5-4951-8506-0314025142fb.png)
</details>

**Comment by Tyriar** [comment]
Just discovered this issue: https://github.com/Microsoft/vscode/issues/27378

**Comment by Tyriar** [comment]
Using typometer, here are the results for xterm.js, with a few changes to get it closer to what we could expect as the upper bound of adopting the webgl renderer:

![image](https://user-images.githubusercontent.com/2193314/197283594-026e2662-afe4-4c95-ac42-9062dba984dd.png)

Those are some pretty nice numbers 🙂, of course the model updates/events in bare xterm.js aren't nearly as costly as in vscode's editor though. The viewport is also quite small in xterm.js' demo which affects the time it takes to fill in the webgl buffers.

Interestingly, those 2 ones that are very low with 8.2-8.3 averages are when I was taking a devtools performance profile. My theory is it's related to performance vs efficiency cores in my i12 CPU. They are also far more consistent during the performance profile:

<img width="580" alt="image" src="https://user-images.githubusercontent.com/2193314/197283884-0d89f8c4-6f1a-468f-b5f9-0a92d52d3048.png">

Compared to the other 3:

<img width="579" alt="image" src="https://user-images.githubusercontent.com/2193314/197283930-0b1ca32a-5f8a-44c4-870e-c8a8ad68b72b.png">

Another thing I noticed are the big spikes that sometimes occur. This is actually why I did the performance profile to have a look at them. Here's an example (edited the screenshot as the iterations part was giant and couldn't be collapsed):

![image](https://user-images.githubusercontent.com/2193314/197284401-ba0c70f0-39d4-47e7-b28d-2a5db674b05c.png)

I can't explain this as the work appears to be done right at the start, like all the other frames. Perhaps it's another application eating more CPU time? Or Chromium doing something and deciding to drop those frames?

**Comment by Tyriar** [comment]
Ok I'm pretty sure it is efficiency vs performance cores. I disabled all efficiency cores in BIOS and I get the good numbers of ~8ms average

**Comment by max-sixty** [comment]
@Tyriar I don't want to add noise to this issue, but I'm excited to see you working on this. I've kept an eye on latency and though I could give some comparisons for context:

I found VSCode quite good — a lower min/max/mean that emacs & vim — at least on a high-spec Macbook Pro. Specifically, from a couple of years ago, typometer got 9.7/27.1/16.8 for min/max/mean; relative to 48/77/60 for vim and 29/48/38 for emacs.

I just reran it a couple of times now and got higher max & means in VSCode 1.72.2: 9.7/66.2/23.1 & 7.4/49.6/21.9. So maybe something has got slightly worse, or maybe the setups a different. Min is still impressively low!

To the extent we get a few ms saving, that's appreciated — I think I can notice a difference when there's more immediate feedback; even small amounts of latency or jitter feel worse (maybe needs a blind test though :) )

**Comment by Tyriar** [comment]
@max-sixty good to hear, I've been trying to squeeze out ms here and there for the past month or so, a lot of the changes are still in review though 🙂

**Comment by Tyriar** [comment]
For the unexplained xterm.js numbers, it appears to be related to extensions, incognito fixes the problem. Perhaps it waits on some extension APIs when not profiling?

**Comment by Tyriar** [comment]
Looking at the spikes in the profiles again I think I can explain it now:

<img width="414" alt="image" src="https://user-images.githubusercontent.com/2193314/197296507-73bc38a8-7ec4-46fd-ab6e-071ad62c6f00.png">

There is a task that does `Compute Intersections` where this happens:

![image](https://user-images.githubusercontent.com/2193314/197296495-11a4aaf9-e79d-444a-a672-b6e1c45932ce.png)

These are for intersection observers which xterm.js [does indeed use](https://github.com/Tyriar/xterm.js/blob/8d4d5bd54ba76a12d25995cc47ab6f9808bfb390/src/browser/services/RenderService.ts#L118-L122). https://developer.chrome.com/en/blog/new-in-devtools-92/#computed-intersections

There are other places where `Compute Intersections` happens that don't have a spike, but all spikes contain a `Compute Intersections`.

**Comment by Tyriar** [comment]
The impact of disabling v-sync and unlocking the frame limit is that it's slightly slower by average but much more variable as you would expect. It doesn't lower the minimum though.

<img width="382" alt="image" src="https://user-images.githubusercontent.com/2193314/197298790-9c792618-0611-47db-b5e5-931a8f8228d3.png">

**Comment by Tyriar** [comment]
Here are some results of measuring latency in VS Code and other similar projects:

![image](https://user-images.githubusercontent.com/2193314/197304156-40b6de50-5451-4025-aa5c-a06aede5e389.png)

Some thoughts:

- It's not a totally fair comparison, since the latency tool is a bit finnicky, some of these could only be measured when the app was either very wide or narrow and certain font sizes, and how large glyphs and the viewport are will likely have a direct relationship with render time. These are not all called out in the image above, I think it was Sublime and all VS Codes that needed to be narrow with a large font size. Inversely the vim tests are in very wide terminals so it's unfair that way too.
- This isn't representative of real world typing as it's period `.` characters into a text file, so it ignores autocomplete for example, or just actual coding in general (varying glyphs, scrolling when typing, wrapping, files with a lot of content, etc.)
- The numbers varied a fair bit in different runs and it's not clear why. For example I did VS Code v1.73.0-insider fairly early on and it was 18ms average, I did another Sublime Text profile just now and it was 14ms average. A more thorough comparison should probably have more individual runs and/or more than 200 characters
- The Insiders ones included my set of extensions
- Electron/Chromium typically had a large maximum as spikes could occur for unknown reasons.
- xterm.js' webgl renderer is super fast which is evident when you conpty from the picture. This used the barebones demo connected to the "fake pty" which basically just prints what you type. So it represents potential of where we could get if the typing did nearly nothing except render, that's unrealistic though as a lot of things need to happen in the editor.

Hardware used:

- CPU: 12th Gen Intel Core i7-12700KF 3.61 GHz (efficiency cores disabled in BIOS)
- GPU: RTX 2070 Super
- Monitor: 240Hz, 3840x2160 resolution, 150% scale

**Comment by Tyriar** [comment]
I'm going to call this done for now, here are the outcomes:

- In depth look at the major things that happen when typing
- Deferred a bunch of stuff when typing
- Setup a telemetry event which will let us track our input latency goals over time and ensure we don't regress
- Learned a lot above
